File: decision_tree/scala/dt.scala
import org.apache.spark.ml.classification.DecisionTreeClassifier
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.ml.feature.VectorAssembler
import org.apache.spark.sql.SparkSession

object DecisionTreeExample {
  def main(args: Array[String]): Unit = {
    // Create SparkSession
    val spark = SparkSession.builder
      .appName("DecisionTreeExample")
      .getOrCreate()

    // Load the dataset (Iris dataset as an example)
    val data = spark.read.format("csv")
      .option("header", "true")
      .option("inferSchema", "true")
      .load("iris.csv")

    // Prepare the data for training
    val assembler = new VectorAssembler()
      .setInputCols(Array("sepal_length", "sepal_width", "petal_length", "petal_width"))
      .setOutputCol("features")
    val dataset = assembler.transform(data)

    // Split the dataset into training and testing sets
    val Array(trainingData, testData) = dataset.randomSplit(Array(0.8, 0.2))

    // Create a decision tree classifier
    val dt = new DecisionTreeClassifier()
      .setLabelCol("species")
      .setFeaturesCol("features")

    // Train the classifier on the training data
    val model = dt.fit(trainingData)

    // Make predictions on the test data
    val predictions = model.transform(testData)

    // Evaluate the model
    val evaluator = new MulticlassClassificationEvaluator()
      .setLabelCol("species")
      .setPredictionCol("prediction")
      .setMetricName("accuracy")
    val accuracy = evaluator.evaluate(predictions)

    println(s"Accuracy: $accuracy")

    // Stop SparkSession
    spark.stop()
  }
}

File: decision_tree/go/dt.go
package main

import (
	"fmt"
	"math/rand"
	"time"

	"github.com/sjwhitworth/golearn/base"
	"github.com/sjwhitworth/golearn/evaluation"
	"github.com/sjwhitworth/golearn/split"
	"github.com/sjwhitworth/golearn/trees"
)

func main() {
	rand.Seed(time.Now().UnixNano())

	// Load the dataset (Iris dataset as an example)
	iris, err := base.ParseCSVToInstances("iris.csv", true)
	if err != nil {
		panic(err)
	}

	// Split the dataset into training and testing sets
	trainData, testData := split.Dataset(iris, &split.KFold{Folds: 2, Shuffle: true}, 0)

	// Create a decision tree classifier
	tree := trees.NewID3DecisionTree(0.6)

	// Train the classifier on the training data
	err = tree.Fit(trainData)
	if err != nil {
		panic(err)
	}

	// Make predictions on the test data
	predictions, err := tree.Predict(testData)
	if err != nil {
		panic(err)
	}

	// Calculate the accuracy of the model
	confusionMat, err := evaluation.GetConfusionMatrix(testData, predictions)
	if err != nil {
		panic(err)
	}
	accuracy := evaluation.GetAccuracy(confusionMat)
	fmt.Printf("Accuracy: %.2f%%\n", accuracy*100)
}
